### ISSUE TITLE
Referenced project ... must have setting "composite": true

### ISSUE THREAD

**Issue by chriscurnow** [nrwl/nx#27292]
### Current Behavior

My `tsconfig` files report an error:  
`Referenced project ... must have setting "composite": true.`

### Expected Behavior

Automatically generated `tsconfig` files should be error free.

### Steps to Reproduce

1. create a new workspace
2. create some projects
3. Everything seems ok at first
4. At some point I will be working on a project and suddenly this error will come up in the `tsconfig` file. I won't be working in the `tsfonfig` file. The error just appears.

### Nx Report

```shell
nx (global)        : 19.4.3
nx                 : 19.5.3
@nx/js             : 19.5.3
@nx/jest           : 19.5.3
@nx/linter         : 19.5.3
@nx/eslint         : 19.5.3
@nx/workspace      : 19.5.3
@nx/angular        : 19.5.3
@nx/devkit         : 19.4.4
@nx/eslint-plugin  : 19.5.3
@nx/playwright     : 19.4.4
@nrwl/tao          : 19.5.3
@nx/web            : 19.5.3
@nx/webpack        : 19.5.3
typescript         : 5.5.4
---------------------------------------
Registered Plugins:
@nx/playwright/plugin
@nx/eslint/plugin
---------------------------------------
Community plugins:
@ngrx/component-store : 18.0.2
@ngrx/effects         : 18.0.2
@ngrx/entity          : 18.0.2
@ngrx/operators       : 18.0.2
@ngrx/router-store    : 18.0.2
@ngrx/schematics      : 18.0.2
@ngrx/signals         : 18.0.2
@ngrx/store           : 18.0.2
---------------------------------------
The following packages should match the installed version of nx
  - @nx/devkit@19.4.4
  - @nrwl/devkit@19.4.4
  - @nx/playwright@19.4.4

To fix this, run `nx migrate nx@19.5.3`
```

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

I realise this report is essentially the same as #20957 but I don't believe it should have been closed.
The error persists in automatically generated tsconfig files. 
This is a bug.
The comment referenced a link which no longer exists: https://github.com/nrwl/nx/issues/20957#issuecomment-1873769215.
@jaysoo commented here https://github.com/nrwl/nx/issues/20957#issuecomment-1883587356, that we were free to add composite:true, but
1. We shouldn't have to
2. This produces errors in many other places.
This should be handled by the generators.

**Comment by arndwestermann** [comment]
I have the same error, this only popped up recently like a month or two ago or so.
You can do a `Restart Extension Host` in VSCode, but if I make any changes to a lib or app, for example add a new file, this error comes back again. I hadn't any trouble with that a while ago, but right now this is really annoying.

**Comment by chriscurnow** [comment]
@arndwestermann, this has to do with the introduction of 'composite' in Typescript.  It's not something I tend to spend a lot of time thinking about. I just want it to work.

I can't find where I found this but this arrangement seems to work.

In `tsconfig.base.json`, include

`"compilerOptions:{
    "declaration": true,
    "composite": true,
    "emitDeclarationOnly": true,
}`

And then in `tsconfig.app.json` or `tsconfig.lib.json` include

`
  "compilerOptions": {

    "emitDecoratorMetadata": true,
    "composite": true,
    "declaration": true,
    "declarationMap": true,
  },
  }`

The problem I have is I think the Nx generators should do this for us.

**Comment by cmrabet** [comment]
I also have this problem and is very annoying. All libs are red flagged.

**Comment by juanchavezlive** [comment]
Same error here since several weeks

**Comment by HenrikTusz** [comment]
Same here

**Comment by Banegasn** [comment]
same problem here any news?

**Comment by rs-r2d2** [comment]
+1

**Comment by dgoerdes** [comment]
I think not using the `composite: true` option is intended by NX.
As setting it to true will make `tsc` to compile all of the referenced packages at once and depends on their code. However in a monorepo setup, as it goes for NX, we want to be able to compile packages individually and not treat everything as one connected thing.

It is annoying the IDE complains about it but it does not affect or causes any other issues than just the IDE complaining.

Please add information or correct me if this is wrong.

I would also love if someone has a solution to keep VSCode silent on this.

**Comment by grigor-minasyan** [comment]
what helped in my case was downgrading typescript to `5.5.3`, the issue only appeared in my case with `5.5.4`

**Comment by alo** [comment]
maybe @juristr can clarify this... 🤷

Thanks in advance

**Comment by Khanon** [comment]
More than two months of this issue and no solution?

**Comment by demirtasdurmus** [comment]
Any progress or ideas on this one?

![Image](https://github.com/user-attachments/assets/7ad4adff-a6cf-4217-9ede-a36c3d64fdbe)

![Image](https://github.com/user-attachments/assets/73888858-dbc1-4b20-bc42-6ae08ba2158f)

**Comment by pwglownia** [comment]
> I think not using the `composite: true` option is intended by NX. As setting it to true will make `tsc` to compile all of the referenced packages at once and depends on their code. However in a monorepo setup, as it goes for NX, we want to be able to compile packages individually and not treat everything as one connected thing.
> 
> It is annoying the IDE complains about it but it does not affect or causes any other issues than just the IDE complaining.
> 
> Please add information or correct me if this is wrong.
> 
> I would also love if someone has a solution to keep VSCode silent on this.

I would simply ts-ignore it, until we find the better solution :)

```
  "references": [
    // @ts-ignore -> https://github.com/nrwl/nx/issues/27292
    {
      "path": "./tsconfig.app.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ],
```

**Comment by KentoMoriwaki** [comment]
I was also troubled by the same error, but it seems to have stopped occurring after I disabled the VSCode Plugin [NX Console](https://marketplace.visualstudio.com/items?itemName=nrwl.angular-console). This plugin might be the cause of the issue.

**Comment by twestendorff** [comment]
We have the same problem. The problem is no longer shown in VSCode when the NX Console version **v18.24.1** is used.

Downgrading may be a workaround, but I agree with the issue creator, the problem shouldn't be there. In the IDE, especially for new project participants, it looks like they did something wrong/something is not working.

**Comment by semanticist21** [comment]
My version is v20.0.5 but the error still occurs.

**Comment by salez** [comment]
this error keeps occuring from time to time, editing the file and saving it again makes the error goes away, but it's pretty annoying

**Comment by nakash2050** [comment]
> I was also troubled by the same error, but it seems to have stopped occurring after I disabled the VSCode Plugin [NX Console](https://marketplace.visualstudio.com/items?itemName=nrwl.angular-console). This plugin might be the cause of the issue.
> 

This worked for me. I miss using Nx console options but could not stand seeing the annoying error. Whole VS Code was lit up.

**Comment by jedihacks** [comment]
Really annoying - happens for us too

**Comment by laszlo-ratesic** [comment]
+1

**Comment by semanticist21** [comment]
https://nx.dev/nx-api/js/documents/typescript-project-references


According to the Doc, they say 'composite' value should be true. I changed the value to true.

**Comment by morgan-wild** [comment]
> https://nx.dev/nx-api/js/documents/typescript-project-references
> 
> According to the Doc, they say 'composite' value should be true. I changed the value to true.

And now, you cannot compile.

**Comment by denys5dev** [comment]
I've been facing the same error in my Nx monorepo with Nx Console and VS Code for the past two years. Interestingly, my colleagues using WebStorm don't encounter this issue. We don't use the composite flag in our configuration.  I'm not entirely sure where the issue lies, or on which side it has to be fixed. it’s more of an annoyance, as it doesn’t affect compilation or functionality.

**Comment by TinekeinTOWN** [comment]
@morgan-wild what solved the compiling issue for me (in my React / Vite app) was:

1. adding  `"composite": true,` to tsconfig.json
2. removing  `"noEmit": true,` from tsconfig.json
3. adding `"composite": true,` to the tsconfig.app.json + tsconfig.node.json reference files 
4. adding '"allowImportingTsExtensions": true,' to the tsconfig.app.json + tsconfig.node.json reference files 
5. changing (or adding) `noEmit` to `emitDeclarationOnly` in the tsconfig.app.json + tsconfig.node.json reference files

**Comment by kolodziejczyks** [comment]
> https://nx.dev/nx-api/js/documents/typescript-project-references
> 
> According to the Doc, they say 'composite' value should be true. I changed the value to true.

This answer resolved issue with errors in VSCode. I just needed to add both properties to `tsconfig.base.json`.

**Comment by namnguyen191** [comment]
This works for me for building Angular libs + using vitest + typescript@5.6.2:
- In `packages/my-lib/tsconfig.json`:
```jsonc
// Note: tsconfig-angular.base.json is just an normal tsconfig file for Angular that I share across multiple libs and apps
{
  "extends": "../../tsconfig-angular.base.json", 
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ]
}
```
- In `packages/my-lib/tsconfig.lib.json`:
```jsonc
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "declaration": true, // THIS IS IMPORTANT
    "composite": true, // THIS IS IMPORTANT
    "declarationMap": true,
    "outDir": "../../../dist/out-tsc",
    "inlineSources": true,
    "types": []
  },
  "exclude": [
    "src/**/*.spec.ts",
    "src/test-setup.ts",
    "jest.config.ts",
    "src/**/*.test.ts",
    "vite.config.ts",
    "vite.config.mts",
    "vitest.config.ts",
    "vitest.config.mts",
    "src/**/*.test.tsx",
    "src/**/*.spec.tsx",
    "src/**/*.test.js",
    "src/**/*.spec.js",
    "src/**/*.test.jsx",
    "src/**/*.spec.jsx",
    "src/test-setup.ts"
  ],
  "include": ["src/**/*.ts"]
}
```
- In `packages/my-lib/tsconfig.spec.json`:
```jsonc
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "declaration": true, // THIS IS IMPORTANT
    "composite": true, // THIS IS IMPORTANT
    "outDir": "../../dist/out-tsc",
    "types": ["vitest/globals", "vitest/importMeta", "vite/client", "node", "vitest"]
  },
  "include": [
    "vite.config.ts",
    "vite.config.mts",
    "vitest.config.ts",
    "vitest.config.mts",
    "src/**/*.test.ts",
    "src/**/*.spec.ts",
    "src/**/*.test.tsx",
    "src/**/*.spec.tsx",
    "src/**/*.test.js",
    "src/**/*.spec.js",
    "src/**/*.test.jsx",
    "src/**/*.spec.jsx",
    "src/**/*.d.ts"
  ],
  "files": ["src/test-setup.ts"],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    }
  ]
}
```
- In `packages/my-lib/tsconfig.lib.prod.json`:
```jsonc
{
  "extends": "./tsconfig.lib.json",
  "compilerOptions": {
    "declaration": false, // THIS IS IMPORTANT
    "composite": false // THIS IS IMPORTANT
  },
  "angularCompilerOptions": {
    "compilationMode": "partial"
  }
}
```
The idea is pretty simple. We need `tsconfig.json` to reference to `tsconfig.lib.json` and `tsconfig.spec.json` for VSCode TS syntax highlighting to work properly. But, we do not want `composite: true` and `declaration: true` when the lib is being build for production ([due to Angular compiler does not support TS incremental build yet](https://github.com/angular/angular/issues/42011)). So, we use a separate `tsconfig.lib.prod.json` that will be ignored by VSCode (since we did not reference it in `tsconfig.json`) but will be used in the build process (it's being used somewhere in `project.json`, under `targets.build.configurations.production.tsConfig`)

**Comment by ThePlenkov** [comment]
any progress on this problem from nx side?

**Comment by ardokirsipuu** [comment]
> https://nx.dev/nx-api/js/documents/typescript-project-references
> 
> According to the Doc, they say 'composite' value should be true. I changed the value to true.

That doc is only for the Nx workspaces that utilize the new [TypeScript Project References](https://nx.dev/blog/announcing-nx-20#typescript-project-references-for-monorepos). However, at this point it's not fully supported yet. From the same blog post:

> Note that we've enabled Project References and Workspaces for the TS preset (i.e. --preset=ts), and we're working on extending this support for all other presets soon (Angular, React, Vue, Node).

Furthermore, I also think this is intended to be false by Nx (for non-TypeScript-Project-References workspaces) for [reasons stated in a previous comment](https://github.com/nrwl/nx/issues/27292#issuecomment-2331270201).

To sum it up, we just need to bare with the IDE errors until somebody manages to address it somehow or until Nx starts fully supporting TypeScript Project References and we manage to migrate our existing workspaces to it.

**Comment by anshul-financeops** [comment]
![Image](https://github.com/user-attachments/assets/a42d0b73-fa39-48cb-909f-2dfd0bff0c04)

I usually keep the NX Console plugin disabled and only enable it when I need to generate apps or libraries. This isn't a perfect solution but helps eliminate the red squiggly lines, ensuring errors only appear for genuine compilation issues. Additionally, I’ve noticed that the cursor responds faster with the plugin disabled.

**Comment by ardokirsipuu** [comment]
The error doesn't come from the plugin. I don't have the plugin enabled, yet I still have these errors.

**Comment by randikasennen** [comment]
I had the same issues mentioned above.
I just upgraded the `typescript` to it's latest version as of now, which is `5.7.3` and the error is gone! 🚀
Give it a shot and comment here...

**Comment by ThePlenkov** [comment]
Hi! Just wanted to tell you that is not the issue with a new setup described here

https://nx.dev/concepts/typescript-project-linking

**Comment by ghost** [comment]
What would be the proper way to get this updated in a existing workspace?
Note: We currently run on "@nx/workspace": "20.8.1".

**Comment by palmthree-studio** [comment]
Hello everyone! 

Maybe it can help for Angular users at least : You can safely empty references:[] from tsconfig, Angular use it's own librairies retrieving system, independent from tsc build, that's the reason why build is successful without "composite:true".
